File: src/python/asynchrony/09_fetch_parallel.py
"""
This script demonstrates how to fetch multiple URLs in parallel using Python's
asyncio and http libraries. The script shows how to create multiple async
tasks, one for each URL to be fetched, and how to run them concurrently using
asyncio.gather(). This approach is beneficial when fetching multiple resources,
as it reduces the total time taken by making requests simultaneously, rather than
sequentially.

The script also includes a synchronous version of the fetch function to compare
the time taken for each approach.
"""

import asyncio
import http.client
from urllib.parse import urlparse
import time

URLS = [
    "https://httpbin.org/get",
    "https://api.github.com",
    "https://www.example.com",
]


def fetch_sync(url):
    url_parsed = urlparse(url)
    conn = http.client.HTTPSConnection(url_parsed.netloc, timeout=10)
    conn.request("GET", url_parsed.path)
    response = conn.getresponse()
    content = response.read()
    print(f"Synchronously fetched content from {url}")
    return content.decode("utf-8")


def fetch_all_sync(urls):
    return [fetch_sync(url) for url in urls]


async def fetch_async(url):
    url_parsed = urlparse(url)
    conn = http.client.HTTPSConnection(url_parsed.netloc, timeout=10)

    async with asyncio.Lock():
        conn.request("GET", url_parsed.path)
        response = conn.getresponse()
        content = response.read()
        print(f"Async fetched content from {url}")
        return content.decode("utf-8")


async def fetch_all_async(urls):
    tasks = [asyncio.create_task(fetch_async(url)) for url in urls]
    return await asyncio.gather(*tasks)


def main():
    start_time_sync = time.time()
    contents_sync = fetch_all_sync(URLS)
    elapsed_time_sync = time.time() - start_time_sync
    print(f"\nSynchronous execution took {elapsed_time_sync} seconds.\n")

    start_time_async = time.time()
    contents_async = asyncio.run(fetch_all_async(URLS))
    elapsed_time_async = time.time() - start_time_async
    print(f"\nAsynchronous execution took {elapsed_time_async} seconds.\n")


if __name__ == "__main__":
    main()

File: src/cpp/asynchrony/09_fetch_parallel.cpp
#include <chrono>
#include <curl/curl.h>
#include <future>
#include <iostream>
#include <string>
#include <vector>

std::vector<std::string> URLS = {
    "https://httpbin.org/get",
    "https://api.github.com",
    "https://www.example.com",
};

size_t WriteCallback(void *contents, size_t size, size_t nmemb, void *userp) {
  ((std::string *)userp)->append((char *)contents, size * nmemb);
  return size * nmemb;
}

std::string fetch_sync(const std::string &url) {
  CURL *curl;
  CURLcode res;
  std::string readBuffer;

  curl_global_init(CURL_GLOBAL_DEFAULT);
  curl = curl_easy_init();
  if (curl) {
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
    res = curl_easy_perform(curl);
    curl_easy_cleanup(curl);
  }
  curl_global_cleanup();

  std::cout << "Synchronously fetched content from " << url << std::endl;
  return readBuffer;
}

std::vector<std::string> fetch_all_sync(const std::vector<std::string> &urls) {
  std::vector<std::string> contents;
  for (const auto &url : urls) {
    contents.push_back(fetch_sync(url));
  }
  return contents;
}

std::string fetch_async(const std::string &url) { return fetch_sync(url); }

std::vector<std::string> fetch_all_async(const std::vector<std::string> &urls) {
  std::vector<std::future<std::string>> futures;
  for (const auto &url : urls) {
    futures.push_back(std::async(std::launch::async, fetch_async, url));
  }

  std::vector<std::string> contents;
  for (auto &future : futures) {
    contents.push_back(future.get());
  }
  return contents;
}

int main() {
  auto start_time_sync = std::chrono::steady_clock::now();
  auto contents_sync = fetch_all_sync(URLS);
  auto elapsed_time_sync =
      std::chrono::duration_cast<std::chrono::milliseconds>(
          std::chrono::steady_clock::now() - start_time_sync)
          .count();
  std::cout << "\nSynchronous execution took " << elapsed_time_sync / 1000.0
            << " seconds.\n";

  auto start_time_async = std::chrono::steady_clock::now();
  auto contents_async = fetch_all_async(URLS);
  auto elapsed_time_async =
      std::chrono::duration_cast<std::chrono::milliseconds>(
          std::chrono::steady_clock::now() - start_time_async)
          .count();
  std::cout << "\nAsynchronous execution took " << elapsed_time_async / 1000.0
            << " seconds.\n";

  return 0;
}
